src/SignalR/clients/csharp/Client/test/FunctionalTests/HubConnectionTests.cs

```diff
@@ -1335,6 +1335,99 @@ public async Task ServerThrowsHubExceptionIfBuildingAsyncEnumeratorIsNotPossible
+                try
+                {
+                    await connection.StartAsync().OrTimeout();
+
+                    var result = connection.InvokeAsync<string>(nameof(TestHub.CallWithUnserializableObject));
+
+                    // The connection should close.
+                    Assert.Null(await closedTcs.Task.OrTimeout());
+
+                    await Assert.ThrowsAsync<TaskCanceledException>(() => result).OrTimeout();
+                }
+                catch (Exception ex)
+                {
+                    LoggerFactory.CreateLogger<HubConnectionTests>().LogError(ex, "{ExceptionType} from test", ex.GetType().FullName);
+                    throw;
+                }
+                finally
+                {
+                    await connection.DisposeAsync().OrTimeout();
+                }
+
+                var errorLog = server.GetLogs().SingleOrDefault(r => r.Write.EventId.Name == "FailedWritingMessage");
+                Assert.NotNull(errorLog);
+                Assert.Contains(exceptionSubstring, errorLog.Write.Exception.Message);
+                Assert.Equal(LogLevel.Error, errorLog.Write.LogLevel);
+            }
+        }
+
+        [Theory]
+        [QuarantinedTest]
+        [MemberData(nameof(HubProtocolsList))]
+        public async Task ServerLogsErrorIfReturnValueCannotBeSerialized(string protocolName)
+        {
+            // Just to help sanity check that the right exception is thrown
+            var exceptionSubstring = protocolName switch
+            {
+                "json" => "A possible object cycle was detected.",
+                "newtonsoft-json" => "A possible object cycle was detected.",
+                "messagepack" => "Failed to serialize Microsoft.AspNetCore.SignalR.Client.FunctionalTests.TestHub+Unserializable value.",
+                var x => throw new Exception($"The test does not have an exception string for the protocol '{x}'!"),
+            };
+
+            var protocol = HubProtocols[protocolName];
+            using (var server = await StartServer<Startup>(write => write.EventId.Name == "FailedWritingMessage"))
+            {
+                var connection = CreateHubConnection(server.Url, "/default", HttpTransportType.LongPolling, protocol, LoggerFactory);
+                var closedTcs = new TaskCompletionSource<Exception>();
+                connection.Closed += (ex) => { closedTcs.TrySetResult(ex); return Task.CompletedTask; };
+                try
+                {
+                    await connection.StartAsync().OrTimeout();
+
+                    var result = connection.InvokeAsync<string>(nameof(TestHub.GetUnserializableObject)).OrTimeout();
+
+                    // The connection should close.
+                    await closedTcs.Task.OrTimeout();
+                }
+                catch (Exception ex)
+                {
+                    LoggerFactory.CreateLogger<HubConnectionTests>().LogError(ex, "{ExceptionType} from test", ex.GetType().FullName);
```

Comment:
Should we be rethrowing here too?

Reply:
Oh whoops, I missed that there were two tests added. I'll fix this one